### PR TITLE
Add support for base_url

### LINKED ISSUE #150: Support base_url
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

Add support for `base_url` per CEP-15, to allow packages to be hosted separately from repodata.json.

A new parameter should add base_url to repodata.json. conda-index will produce but not consume the new metadata.

### Why is this needed?

_No response_

### What should happen?

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dholth**
### Description

Add support for base_url. Fix #150 

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

File: conda_index/cli/__init__.py [hunk 1]
```diff
@@ -83,6 +83,14 @@
         will keep python 3.8.X and 3.9.Y in the current_index.json, instead of only the very latest python version.
         """,
 )
+@click.option(
+    "--base-url",
+    help="""
+        If packages should be served separately from repodata.json, URL of the
+        directory tree holding packages. Generates repodata_version=2 which is
+        incompatible with existing versions of conda.
+        """,
```
Comment (dholth) [inline conda_index/cli/__init__.py:92]: Would like to find less awkward wording here.

**Comment by dholth** [comment]
Should we add base_url to index.html also?

File: conda_index/index/__init__.py [hunk 2]
```diff
@@ -499,8 +500,10 @@ def __init__(
         write_zst=False,
         write_run_exports=False,
         compact_json=True,
+        *,
```
Comment (dholth) [inline conda_index/index/__init__.py:503]: make newer arguments keyword-only and not passable by position

File: conda_index/cli/__init__.py (hunk 1 above)
Comment (jezdez) [reply conda_index/cli/__init__.py:92]: I wonder if we should introduce a `--repodata-version=N` option, and `--base-url` basically implies this, when passed? That make it more explicit what is being generated and help when users migrate?

That also reminds me that we need to add a schema for it to https://github.com/conda/schemas.

File: conda_index/cli/__init__.py [hunk 3]
```diff
@@ -83,6 +83,14 @@
         will keep python 3.8.X and 3.9.Y in the current_index.json, instead of only the very latest python version.
         """,
 )
+@click.option(
+    "--base-url",
+    help="""
+        If packages should be served separately from repodata.json, URL of the
+        directory tree holding packages. Generates repodata_version=2 which is
+        incompatible with existing versions of conda.
```
Comment (jezdez) [inline conda_index/cli/__init__.py:91]:
```suggestion
        directory tree holding packages. Generates repodata.json with
        repodata_version=2 which is support in conda 24.5.0 or later.
```

Comment (jezdez) [reply conda_index/cli/__init__.py:91]: https://github.com/conda/conda/issues/13137 is now fixed

File: conda_index/index/__init__.py [hunk 4]
```diff
@@ -748,6 +752,11 @@ def index_subdir(self, subdir, verbose=False, progress=False):
             "removed": [],  # can be added by patch/hotfix process
         }
 
+        if self.base_url:
+            # per https://github.com/conda-incubator/ceps/blob/main/cep-15.md
+            new_repodata["info"]["base_url"] = f"{self.base_url.rstrip('/')}/{subdir}/"
+            new_repodata["repodata_version"] = 2
```
Comment (jezdez) [inline conda_index/index/__init__.py:758]: Assuming you'd be okay to add a --repodata-version option (which defaults to 1), we could add the repodata_version to all repodata generated?

```suggestion
        new_repodata["repodata_version"] = self.repodata_version
        if self.repodata_version == 2:
            if not self.base_url:
                raise ValueError(
                    "Repodata version 2 was request but no base URL specified."
                )
            # per https://github.com/conda-incubator/ceps/blob/main/cep-15.md
            new_repodata["info"]["base_url"] = f"{self.base_url.rstrip('/')}/{subdir}/"
```

Comment (dholth) [reply conda_index/index/__init__.py:758]: I was thinking along the lines of automatically choosing "the smallest repodata version number supporting features requested"

**Review by jezdez — APPROVED** [review]

**Pull request merged**